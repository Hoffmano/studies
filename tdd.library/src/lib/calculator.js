const sum = (number1, number2) => {
  number1 = parseInt(number1)
  number2 = parseInt(number2)

  if (Number.isNaN(number1) || Number.isNaN(number2))
    throw new Error()

  return number1 + number2
}

export { sum }
