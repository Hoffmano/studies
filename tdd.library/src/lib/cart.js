import { gt, equals } from 'ramda'
import Dinero from 'dinero.js'
const Money = Dinero

Money.defaultCurrency = 'BRL'
Money.defaultPrecision = 2
Money.defaultAmount = 0

const getTotal = cart =>
  cart.reduce(
    (accumulator, item) =>
      accumulator.add(Money({ amount: item.quantity * item.product.price })),
    Money(),
  )

const addItem = (cart, item, quantity = 1) => {
  if (quantity) item = { ...item, quantity: quantity }

  const foundItemIndex = cart.findIndex(cartItem =>
    equals(cartItem.product, item.product),
  )

  if (gt(foundItemIndex, -1)) {
    cart[foundItemIndex].quantity += item.quantity
  } else {
    cart.push(item)
  }

  return cart
}

const updateQuantity = (cart, item, quantity) => {
  item = { ...item, quantity }

  const foundItemIndex = cart.findIndex(cartItem =>
    equals(cartItem.product, item.product),
  )

  if (gt(foundItemIndex, -1)) {
    cart[foundItemIndex].quantity += item.quantity

    if (cart[foundItemIndex].quantity <= 0) {
      cart.splice(foundItemIndex, 1)
    }
  } else {
    cart.push(item)
  }

  return cart
}

const summary = cart => {
  const total = getTotal(cart).getAmount()
  const checkout = { items: cart, total }
  return checkout
}

const checkout = cart => {
  const checkout = summary(cart)
  return { updatedCart: [], checkoutCart: checkout }
}

export { getTotal, addItem, updateQuantity, summary, checkout }
