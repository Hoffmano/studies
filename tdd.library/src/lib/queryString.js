const keyValueToString = ([key, value]) => {
  if (typeof value === 'object' && !Array.isArray(value)) throw new Error()
  return `${key}=${value}`
}

const queryString = object =>
  Object.entries(object).map(keyValueToString).join('&')

const parse = queryString =>
  Object.fromEntries(
    queryString.split('&').map(attribute => {
      let [key, value] = attribute.split('=')
      if (value.indexOf(',') !== -1) value = value.split(',')
      return [key, value]
    }),
  )

module.exports = { queryString, parse }
