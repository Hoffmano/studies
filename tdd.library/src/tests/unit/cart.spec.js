import { getTotal, updateQuantity, checkout, summary } from '../../lib/cart'

let sneakers = {
  product: {
    title: 'TÃªnis',
    price: 5000,
  },
  quantity: 1,
}

let shirt = {
  product: {
    title: 'Shirt',
    price: 3000,
  },
  quantity: 1,
}

describe('CART', () => {
  it('should return 0 when getTotal() is called in a newly created cart', () => {
    let cart = []

    expect(getTotal(cart).getAmount()).toEqual(0)
  })

  it('should add an item to cart', () => {
    let cart = []

    cart = updateQuantity(cart, sneakers, 1)

    expect(cart).toEqual([sneakers])
  })

  it('should allow just one item of each product in the cart', () => {
    let cart = []

    cart = updateQuantity(cart, sneakers, 1)
    cart = updateQuantity(cart, sneakers, 2)

    expect(cart).toEqual([{ ...sneakers, quantity: 3 }])
  })

  it('should multiply quantity by price and return the total amount', () => {
    let cart = []

    cart = updateQuantity(cart, sneakers, 2)

    expect(getTotal(cart).getAmount()).toEqual(10000)
  })

  it('should update total amount in cart', () => {
    let cart = []

    expect(getTotal(cart).getAmount()).toEqual(0)

    cart = updateQuantity(cart, shirt, 1)
    expect(getTotal(cart).getAmount()).toEqual(3000)

    cart = updateQuantity(cart, sneakers, 1)
    expect(getTotal(cart).getAmount()).toEqual(8000)
  })

  it('should remove some item from cart', () => {
    let cart = []

    cart = updateQuantity(cart, shirt, 1)
    cart = updateQuantity(cart, shirt, -1)
    expect(cart).toEqual([])
  })

  it('should remove some item from cart', () => {
    let cart = []

    cart = updateQuantity(cart, shirt, 1)
    cart = updateQuantity(cart, sneakers, 11)

    expect(summary(cart)).toMatchSnapshot()
  })

  it('should cleanup cart', () => {
    let cart = []

    cart = updateQuantity(cart, shirt, 1)
    cart = updateQuantity(cart, sneakers, 11)

    const { updatedCart, checkoutCart } = checkout(cart)
    
    expect(getTotal(updatedCart).getAmount()).toEqual(0)
    expect(checkoutCart).toMatchSnapshot()
  })
})
