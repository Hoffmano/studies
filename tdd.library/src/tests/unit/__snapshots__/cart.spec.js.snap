// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CART should cleanup cart 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 3000,
        "title": "Shirt",
      },
      "quantity": 1,
    },
    Object {
      "product": Object {
        "price": 5000,
        "title": "Tênis",
      },
      "quantity": 11,
    },
  ],
  "total": 58000,
}
`;

exports[`CART should remove some item from cart 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 3000,
        "title": "Shirt",
      },
      "quantity": 1,
    },
    Object {
      "product": Object {
        "price": 5000,
        "title": "Tênis",
      },
      "quantity": 11,
    },
  ],
  "total": 58000,
}
`;
